sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,488 @@
+name, except the following cases:
+- Declarations with `__` prefix in demangled name are not recorded as imported
+  functions
+  - Declarations with `__spirv_*` prefix should not be recorded as dependencies
+  since they represent SPIR-V operations and will be transformed to SPIR-V
+  instructions during LLVM->SPIR-V translation.
+- Based on some attributes (which could be defined later) we may want to avoid
+  listing some functions as imported ones
+  - This is needed to have possibility to call device-specific builtins not
+    starting with `__` by forward-declaring them in DPC++ code
+
+**NOTE**: If device code split is enabled, imports collection is performed after
+split and it is performed on separated images.
+
+All collected information is attached to a device image via properties
+mechanism.
+
+Each device image is supplied with an array of property sets:
+```
+struct pi_device_binary_struct {
+...
+  // Array of property sets
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+
+```
+Each property set is represented by the following struct:
+```
+// Named array of properties.
+struct _pi_device_binary_property_set_struct {
+  char *Name;                                // the name
+  pi_device_binary_property PropertiesBegin; // array start
+  pi_device_binary_property PropertiesEnd;   // array end
+};
+```
+It contains name of property set and array of properties. Each property is
+represented by the following struct:
+```
+struct _pi_device_binary_property_struct {
+  char *Name;       // null-terminated property name
+  void *ValAddr;    // address of property value
+  uint32_t Type;    // _pi_property_type
+  uint64_t ValSize; // size of property value in bytes
+};
+```
+
+List of imported symbols is represented as a single property set with name
+`SYCL/imported symbols` recorded in the `Name` field of property set.
+Each property in this set holds name of the particular imported symbol recorded
+in the `Name` field of the property.
+List of exported symbols is represented in the same way, except the
+corresponding set has the name `SYCL/exported symbols`.
+
+### DPC++ runtime changes
+
+DPC++ RT performs *device images collection* task by grouping all device
+images required to execute a kernel based on the list of exports/imports and
+links them together using PI API.
+
+#### Device images collection
+
+DPC++ Runtime class named ProgramManager stores device images using following
+data structure:
+```
+/// Keeps all available device executable images added via \ref addImages.
+/// Organizes the images as a map from a kernel set id to the vector of images
+/// containing kernels from that set.
+/// Access must be guarded by the \ref Sync::getGlobalLock()
+std::unordered_map<SymbolSetId,
+                   std::unique_ptr<std::vector<RTDeviceBinaryImageUPtr>>>
+    m_DeviceImages;
+
+using StrToKSIdMap = std::unordered_map<string_class, SymbolSetId>;
```

Comment:
I think use-case when kernels have the same name is not really legal. SYCL 2020 spec says "In the case of the shared-source compilation model, the kernels have to be uniquely identified by both host and device compiler.". So if two kernels (even if they are in different objects) have the same name, they don't really uniquely identified, right?

Reply:
Let's have the following use-case:
```c++
// Application
queue Q;
Lib1Func(Q);
Lib2Func(Q);

// Lib1
void Lib1Func(queue &Q) {
  Q.sbumit([](handle &H) {
    H.parallel_for<class Calculate>(...);
  });
}

// Lib2
void Lib2Func(queue &Q) {
  Q.sbumit([](handle &H) {
    H.parallel_for<class Calculate>(...);
  });
}

```

Is this use-case handled by SYCL 2020 spec?

Reply:
I'm not sure that SYCL 2020 (or other) spec say something about such use cases, especially when libraries are involved. However, looking at this comment https://github.com/intel/llvm/pull/3210#discussion_r612066000 , I think this case is illegal. @gmlueck , could you please confirm?

Reply:
I believe this is illegal because the spec says this:

> In the case of the shared-source compilation model, the kernels have to be **uniquely identified** by both host and device compiler. 

The spec could certainly be more clear about this, though.  I'll propose updated wording to the SYCL 2020 spec clarifying that a kernel name must be unique across the entire application.

Note that it **is** legal for two translation units (or two libraries) to call the **same** kernel, in which case both instances of the kernel will have the same name.  For example, consider a named kernel object that is called from two different libraries:

```
class MyKernel {
 public:
  void operator()(item<1> it) const {
    /* kernel code */
  }
};
```

In such a case, both libraries will have a definition of the kernel named `MyKernel`.  The design will need to support cases like this.